File: Crypto4o/crypto4o.py
#!/usr/bin/python

import sys

# ----------------------------------------------------------------------------
# Crypto4o functions start here
# ----------------------------------------------------------------------------

class glfsr:
    def __init__(self, polynom, initial_value):
        self.polynom = polynom | 1
        self.data = initial_value
        tmp = polynom

        self.mask = 1

        while tmp != 0:
            if tmp & self.mask != 0:
                tmp = tmp ^ self.mask;

            if tmp == 0:
                break

            self.mask = self.mask << 1

    def next_state(self):
        self.data = self.data << 1

        retval = 0

        if self.data & self.mask != 0:
            retval = 1
            self.data = self.data ^ self.polynom

        return retval

class sprng:
    def __init__(self, polynom_d, init_value_d, polynom_c, init_value_c):
        self.glfsr_d = glfsr(polynom_d, init_value_d)
        self.glfsr_c = glfsr(polynom_c, init_value_c)

    def next_byte(self):
        byte = 0
        bitpos = 7

        while 1 == 1:
            bit_d = self.glfsr_d.next_state()
            bit_c = self.glfsr_c.next_state()

            if bit_c != 0:
                bit_r = bit_d
                byte = byte | (bit_r << bitpos)

                bitpos = bitpos - 1

                if bitpos < 0:
                    break

        return byte

# ----------------------------------------------------------------------------
# Crypto4o functions end here
# ----------------------------------------------------------------------------

def main():
    prng = sprng(int(sys.argv[3], 16), int(sys.argv[4], 16), 
        int(sys.argv[5], 16), int(sys.argv[6], 16))

    print "GLFSR D0: using polynom 0x%X, initial value: 0x%X." % (int(sys.argv[3], 16), int(sys.argv[4], 16))
    print "GLFSR C0: using polynom 0x%X, initial value: 0x%X." % (int(sys.argv[5], 16), int(sys.argv[6], 16))

    f = open(sys.argv[1], "rb")
    g = open(sys.argv[2], "wb")

    while 1 == 1:
        input_ch = f.read(1)

        if input_ch == "":
            break

        random_ch = prng.next_byte() & 0xff
        g.write(chr(ord(input_ch) ^ random_ch))

    f.close()
    g.close()

main()

File: Crypto4o/crypto4o.c
#include <stdio.h>
#include <stdlib.h>
#include <glfsr.h>
#include <crypto4o.h>

unsigned int bitcounter = 0;

t_sprng sprng;

void sprng_init(t_sprng *sprng, t_glfsr_data polynom_d, t_glfsr_data init_value_d, t_glfsr_data polynom_c, t_glfsr_data init_value_c) {
    glfsr_init(&(sprng->glfsr_d0), polynom_d, init_value_d ? init_value_d : 1);
    glfsr_init(&(sprng->glfsr_c0), polynom_c, init_value_c ? init_value_c : 1);
}

unsigned char sprng_next_byte(t_sprng *sprng) {
    unsigned char byte = 0;
    char bitpos = 7;
    unsigned char bit0, bitc0, bitr;

    do {
        bit0 = glfsr_next_state(&(sprng->glfsr_d0));
        bitc0 = glfsr_next_state(&(sprng->glfsr_c0));

#ifdef __DEBUG__
        if (sprng->glfsr_d0.data == sprng->init_value_d) {
            printf("GLFSR D0 overflow at %d.\n", bitcounter);
        }

        if (sprng->glfsr_c0.data == sprng->init_value_c) {
            printf("GLFSR C0 overflow at %d.\n", bitcounter);
        }

        if ((sprng->glfsr_d0.data == sprng->init_value_d) && (sprng->glfsr_c0.data == sprng->init_value_c)) {
            printf("**** GLFSR overflow at %d. ****\n", bitcounter);
        }

        printf("Next bits (%d): %d, %d\t= ", bitpos, bit0, bitc0);
#endif

        if (bitc0) {
            bitr = bit0;

#ifdef __DEBUG__
            printf("%d\n", bitr);
#endif

            byte |= bitr << bitpos;
            bitpos--;
        } else {
#ifdef __DEBUG__
            printf("dropped\n");
#endif
        }

        bitcounter++;
    } while (bitpos >= 0);

    return byte;
}
